### PR TITLE
Configurable run limit on products.

### LINKED ISSUE #1410: new configuration variable: introduce a per product configurable run limit
Today we have a globally configurable limit for the number of runs.
Introduce a similar limit on the product level. The variable should only be modifiable by the product admin.

### PR DESCRIPTION
**Pull request by csordasmarton**
> Closes #1410

Run limit can be configured on each product separately.

### CODE REVIEW

**Comment by whisperity** [comment]
Please extensively check this patch for backwards compatibility. I'm not sure how Thrift handles the expansion of structs. What happens if older client tries to send the configuration struct?

**Comment by csordasmarton** [comment]
@whisperity We did this kind of expansion for a couple of times (e.g.: #1166, #968). I've also checked this server version with an older client and it works fine (add/remove/list products from command line).

**Comment by whisperity** [comment]
@csordasmarton Alright then. Was a first glance exclamation, I didn't recall these improvements.

File: api/v6/products.thrift [hunk 1]
```diff
@@ -40,6 +41,7 @@ struct Product {
   6: bool              accessible,        // Indicates whether the current user can access this product.
   7: bool              administrating     // Indicates that the current user can administrate the product.
```
Comment (whisperity) [inline api/v6/products.thrift]: Comma missing

File: api/v6/products.thrift [hunk 2]
```diff
@@ -40,6 +41,7 @@ struct Product {
...
   8: shared.DBStatus   databaseStatus     // Indicates the database backend status.
```
Comment (whisperity) [inline api/v6/products.thrift]: Comma missing

File: api/v6/products.thrift [hunk 3]
```diff
@@ -40,6 +41,7 @@ struct Product {
...
+  9: i64               runLimit           // Number of allowed runs for this product.
```
Comment (whisperity) [inline api/v6/products.thrift]: Most of the other structs have `,` after the last element so adding a new member does not make the previous line modified.

File: api/v6/products.thrift [hunk 4]
```diff
@@ -25,7 +25,8 @@ struct ProductConfiguration {
   2:          string             endpoint,
   3:          string             displayedName_b64,
   4:          string             description_b64,
-  5: optional DatabaseConnection connection
+  5: optional DatabaseConnection connection,
+  6:          i64                runLimit
```
Comment (whisperity) [inline api/v6/products.thrift]: Most of the other structs have `,` after the last element so adding a new member does not make the previous line modified.

File: config_db_migrate/versions/bb5278995f41_run_limitation_for_products.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+down_revision = '150800b30447'
+branch_labels = None
+depends_on = None
+
+from alembic import op
+import sqlalchemy as sa
+
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
```
Comment (whisperity) [inline config_db_migrate/versions/bb5278995f41_run_limitation_for_products.py]: Once the autogenerated operations had been revised to work, these autogenerated comments should be removed.

File: www/scripts/productlist/ProductSettingsView.js [hunk 6]
```diff
@@ -69,7 +69,7 @@ function (declare, domAttr, domClass, domConstruct, Dialog, Button,
     0 : ['submit'],
     /* PRODUCT_ADMIN */ 1 : ['name', 'description'],
     /* SUPERUSER */     2 : ['endpoint', 'dbengine', 'dbhost', 'dbport',
-                             'dbuser', 'dbpass', 'dbname']
+                             'dbuser', 'dbpass', 'dbname', 'runlimit']
```
Comment (whisperity) [inline www/scripts/productlist/ProductSettingsView.js]: #1410 states that `product admin`s should be able to change this variable too, not just superusers.

File: www/scripts/productlist/ProductSettingsView.js [hunk 7]
```diff
@@ -267,6 +280,7 @@ function (declare, domAttr, domClass, domConstruct, Dialog, Button,
 
       var product = new CC_PROD_OBJECTS.ProductConfiguration({
         endpoint: args['endpoint'],
+        runLimit: args['runlimit'] ? args['runlimit'] : -1,
```
Comment (whisperity) [inline www/scripts/productlist/ProductSettingsView.js]: How does `-1` harmonise with the logic behind this field? Can't we set this to `null` and send a `None` into Python?

File: www/scripts/productlist/ProductSettingsView.js (hunk 6 above)
Comment (gyorb) [reply www/scripts/productlist/ProductSettingsView.js]: @whisperity If I remember it correctly we discussed that only the superuser should  have the right to change this.

Comment (whisperity) [reply www/scripts/productlist/ProductSettingsView.js]: Sure. :wink: I don't think I was there on that discussion. But it sounds good that it's a superuser thing.

**Comment by csordasmarton** [comment]
@gyorb I couldn't reproduce the above problem. I started a new web server with an empty workspace, updated the run limit but I didn't get any error. Could you please check your configuration database? Is it valid?

**Comment by gyorb** [comment]
the server side error is not connected to these changes

**Pull request merged**